sw/host/opentitanlib/src/transport/hyperdebug/i2c.rs

```diff
@@ -0,0 +1,234 @@
+
+/// Wire format of USB packet to request a short I2C transaction
+/// (receiving at most 127 bytes).
+#[derive(AsBytes, FromBytes, Debug)]
+#[repr(C)]
+struct CmdTransferShort {
+    port: u8,
+    addr: u8,
+    write_count: u8,
+    read_count: u8,
+    data: [u8; USB_MAX_SIZE - 4],
+}
+impl CmdTransferShort {
+    fn new() -> Self {
+        Self {
+            port: 0,
+            addr: 0,
+            write_count: 0,
+            read_count: 0,
+            data: [0; USB_MAX_SIZE - 4],
+        }
+    }
+}
+
+/// Wire format of USB packet to request a long I2C transaction
+/// (receiving up to 32767 bytes).
+#[derive(AsBytes, FromBytes, Debug)]
+#[repr(C)]
+struct CmdTransferLong {
+    port: u8,
+    addr: u8,
+    write_count: u8,
+    read_count: u8,
+    read_count1: u8,
+    reserved: u8,
+    data: [u8; USB_MAX_SIZE - 6],
+}
+impl CmdTransferLong {
+    fn new() -> Self {
+        Self {
+            port: 0,
+            addr: 0,
+            write_count: 0,
+            read_count: 0,
+            read_count1: 0,
+            reserved: 0,
+            data: [0; USB_MAX_SIZE - 6],
+        }
+    }
+}
+
+/// Wire format of USB packet containing I2C transaction response.
+#[derive(AsBytes, FromBytes, Debug)]
+#[repr(C)]
+struct RspTransfer {
+    status_code: u16,
+    reserved: u16,
+    data: [u8; USB_MAX_SIZE - 4],
+}
+impl RspTransfer {
+    fn new() -> Self {
+        Self {
+            status_code: 0,
+            reserved: 0,
+            data: [0; USB_MAX_SIZE - 4],
+        }
+    }
+}
+
+impl HyperdebugI2cTarget {
+    pub fn open(hyperdebug: &Hyperdebug, idx: u8) -> Result<Self> {
+        let mut usb_handle = hyperdebug.inner.usb_device.borrow_mut();
+
+        // Exclusively claim I2C interface, preparing for bulk transfers.
+        usb_handle.claim_interface(hyperdebug.i2c_interface.interface)?;
+
+        Ok(Self {
+            inner: Rc::clone(&hyperdebug.inner),
+            interface: hyperdebug.i2c_interface,
+            bus_idx: idx,
+            max_read_size: 0x8000 as usize,
+            max_write_size: 0x1000 as usize,
+        })
+    }
+
+    /// Transmit data for a single I2C operation, using one or more USB packets.
+    fn transmit_then_receive(&self, addr: u8, wbuf: &[u8], rbuf: &mut [u8]) -> Result<()> {
+        ensure!(
+            rbuf.len() < self.max_read_size,
+            I2cError::InvalidDataLength(rbuf.len())
+        );
+        ensure!(
+            wbuf.len() < self.max_write_size,
+            I2cError::InvalidDataLength(wbuf.len())
+        );
+        let mut index = if rbuf.len() < 128 {
+            // Short format header
+            let mut req = CmdTransferShort::new();
+            req.port = self.bus_idx | (((wbuf.len() & 0x0F00) >> 4) as u8);
+            req.addr = addr;
+            req.write_count = (wbuf.len() & 0x00FF) as u8;
+            req.read_count = rbuf.len() as u8;
+            let databytes = std::cmp::min(USB_MAX_SIZE - 4, wbuf.len());
```

Comment:
Consider a `use std::cmp`; and writing this as `cmp::min`.

Reply:
Done